pkg/auth/requests/impersonate.go

```diff
@@ -53,23 +58,41 @@ func (h *impersonatingAuth) Authenticate(req *http.Request) (k8sUser.Info, bool,
 			if err != nil {
 				return nil, false, err
 			} else if !canDo {
-				return nil, false, errors.New("not allowed to impersonate")
+				return nil, false, errors.New("not allowed to impersonate user")
 			}
 			impersonateUser = true
 		}
 
-		if len(reqGroup) > 0 && !groupsEqual(reqGroup, groups) {
-			canDo, err := h.sar.UserCanImpersonateGroups(req, user, reqGroup)
+		if len(reqGroup) > 0 {
+			for _, g := range reqGroup {
+				if slices.Contains(groups, g) {
+					//user belongs to the group they are trying to impersonate
+					continue
+				}
+				canDo, err := h.sar.UserCanImpersonateGroup(req, user, g)
+				if err != nil {
+					return nil, false, err
+				} else if !canDo {
+					return nil, false, errors.New("not allowed to impersonate group")
+				}
+				impersonateGroup = true
+			}
+		}
+
+		if reqExtras != nil && len(reqExtras) > 0 {
+			canDo, err := h.sar.UserCanImpersonateExtras(req, user, reqExtras)
 			if err != nil {
 				return nil, false, err
 			} else if !canDo {
-				return nil, false, errors.New("not allowed to impersonate")
+				return nil, false, errors.New("not allowed to impersonate extras")
 			}
-			impersonateGroup = true
+			impersonateExtras = true
 		}
 	}
 
-	if impersonateUser || impersonateGroup {
+	var extras map[string][]string
+
+	if impersonateUser {
```

Comment:
I think the behavior here changed. Now the `extras = userInfo.GetExtra()` is populated only if the `impersonateUser` is false. In the previous version this was always populated. What about the group impersonation? If the `impersonateGroup` can be true only if the `impersonateUser` is true then it should be fine.

Also the next line with the same `if` is redundant:

```go
if impersonateUser {
	if impersonateUser {
		user = reqUser
	}
}
```

probably it can be simplified with something like:

```go
if impersonateUser {
	user = reqUser
	groups = nil

	if impersonateGroup {
		groups = reqGroup
	}
	groups = append(groups, k8sUser.AllAuthenticated)

	if impersonateExtras {
		extras = reqExtras
	}
}
```

and if the `k8sUser.AllAuthenticated` doesn't need to be at the end:

```go
if impersonateUser {
	user = reqUser
	groups = []string{k8sUser.AllAuthenticated}

	if impersonateGroup {
		groups = append(groups, reqGroup...)
	}
	if impersonateExtras {
		extras = reqExtras
	}
}
```

Reply:
> I think the behavior here changed. Now the extras = userInfo.GetExtra() is populated only if the impersonateUser is false. In the previous version this was always populated

True. That was actually the bug. Extras were always populated with the info from the user doing the request (userInfo.GetExtra()). Because of that we were trying to impersonate the user using the extras from the user doing the request not the impersonated user, which was causing the following error:
```
Error from server (Forbidden): userextras.authentication.k8s.io "local://u-v7dxk" is forbidden: User "system:serviceaccount:cattle-impersonation-system:cattle-impersonation-u-s984r" cannot impersonate resource "userextras/principalid" in API group "authentication.k8s.io" at the cluster scope
```